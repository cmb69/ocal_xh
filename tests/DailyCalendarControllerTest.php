<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Ocal_XH.
 *
 * Ocal_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ocal_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ocal_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Ocal;

use DateTimeImmutable;
use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use XH\CSRFProtection as CsrfProtector;

class DailyCalendarControllerTest extends TestCase
{
    /** @var DailyCalendarController */
    private $sut;

    /** @var ListService&MockObject */
    private $listService;

    public function setUp(): void
    {
        global $plugin_cf;

        $_SERVER['QUERY_STRING'] = "";
        $csrfProtector = $this->createStub(CsrfProtector::class);
        $csrfProtector->method('tokenInput')->willReturn(
            '<input type="hidden" name="xh_csrf_token" value="dcfff515ebf5bd421d5a0777afc6358b">'
        );
        $plugin_cf = XH_includeVar("./config/config.php", 'plugin_cf');
        $config = $plugin_cf['ocal'];
        $plugin_tx = XH_includeVar("./languages/en.php", 'plugin_tx');
        $lang = $plugin_tx['ocal'];
        $now = new DateTimeImmutable("2023-07-02");
        $this->listService = $this->createStub(ListService::class);
        $db = $this->createStub(Db::class);
        $db->method('findOccupancy')->willReturn(new DailyOccupancy("test-daily", 3));
        $this->sut = new DailyCalendarController(
            "/",
            "./",
            $csrfProtector,
            $config,
            $lang,
            $now,
            $this->listService,
            $db,
            true,
            "test-daily",
            1
        );
    }

    public function testDefaultActionRendersCalendar(): void
    {
        $response = $this->sut->defaultAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testDefaultActionHandlesAjaxRequest(): void
    {
        $_SERVER['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest';
        $_GET = ['ocal_name' => "test-daily"];
        $response = $this->sut->defaultAction("test-daily", 1);
        $this->assertEquals("text/html", $response->contentType());
        Approvals::verifyHtml($response->output());
    }

    public function testListActionRendersListWithoutEntries(): void
    {
        $response = $this->sut->listAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testListActionRendersListWithAnEntry(): void
    {
        $this->listService->method('getDailyList')->willReturn([
            (object) ['range' => "9.", 'state' => "2", 'label' => "available"],
        ]);
        $response = $this->sut->listAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testListActionHandlesAjaxRequest(): void
    {
        $_SERVER['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest';
        $_GET = ['ocal_name' => "test-daily"];
        $response = $this->sut->listAction("test-daily", 1);
        $this->assertEquals("text/html", $response->contentType());
        Approvals::verifyHtml($response->output());
    }

    public function testSaveActionReturnsEmptyResponseNameIsMissing(): void
    {
        $response = $this->sut->saveAction("test-daily", 1);
        $this->assertEquals("", $response->output());
    }

    public function testSaveActionReportsSuccess(): void
    {
        $_GET = ['ocal_name' => "test-daily"];
        $_POST = ['ocal_states' => json_encode(['2023-02' => array_fill(0, 27, "1")])];
        $response = $this->sut->saveAction("test-daily", 1);
        $this->assertEquals('<p class="xh_success">Successfully saved.</p>', $response->output());
    }
}
