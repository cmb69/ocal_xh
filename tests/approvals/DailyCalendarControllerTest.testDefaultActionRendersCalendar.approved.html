<!-- daily-calendars -->
<div class="ocal_calendars" data-name="test-daily">
  <!-- mode-link -->
<p class="ocal_mode">
  <a class="ocal_button" href="/?ocal_action=list">List view</a>
</p>
  <!-- statusbar -->
<div class="ocal_loaderbar"><img src="./images/ajax-loader-bar.gif" alt="loading"></div>
<div class="ocal_statusbar"></div>
  <!-- daily-calendar -->
<table class="ocal_calendar" data-ocal_date="2023-07">
  <thead>
    <tr>
      <th colspan="7">July 2023</th>
    </tr>
    <tr>
      <th>M</th>
      <th>T</th>
      <th>W</th>
      <th>T</th>
      <th>F</th>
      <th>S</th>
      <th>S</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td class="ocal_state " data-ocal_state="0" title="">1</td>
      <td class="ocal_state " data-ocal_state="0" title="">2</td>
    </tr>
    <tr>
      <td class="ocal_state " data-ocal_state="0" title="">3</td>
      <td class="ocal_state " data-ocal_state="0" title="">4</td>
      <td class="ocal_state " data-ocal_state="0" title="">5</td>
      <td class="ocal_state " data-ocal_state="0" title="">6</td>
      <td class="ocal_state " data-ocal_state="0" title="">7</td>
      <td class="ocal_state " data-ocal_state="0" title="">8</td>
      <td class="ocal_state " data-ocal_state="0" title="">9</td>
    </tr>
    <tr>
      <td class="ocal_state " data-ocal_state="0" title="">10</td>
      <td class="ocal_state " data-ocal_state="0" title="">11</td>
      <td class="ocal_state " data-ocal_state="0" title="">12</td>
      <td class="ocal_state " data-ocal_state="0" title="">13</td>
      <td class="ocal_state " data-ocal_state="0" title="">14</td>
      <td class="ocal_state " data-ocal_state="0" title="">15</td>
      <td class="ocal_state " data-ocal_state="0" title="">16</td>
    </tr>
    <tr>
      <td class="ocal_state " data-ocal_state="0" title="">17</td>
      <td class="ocal_state " data-ocal_state="0" title="">18</td>
      <td class="ocal_state " data-ocal_state="0" title="">19</td>
      <td class="ocal_state " data-ocal_state="0" title="">20</td>
      <td class="ocal_state " data-ocal_state="0" title="">21</td>
      <td class="ocal_state " data-ocal_state="0" title="">22</td>
      <td class="ocal_state " data-ocal_state="0" title="">23</td>
    </tr>
    <tr>
      <td class="ocal_state " data-ocal_state="0" title="">24</td>
      <td class="ocal_state " data-ocal_state="0" title="">25</td>
      <td class="ocal_state " data-ocal_state="0" title="">26</td>
      <td class="ocal_state " data-ocal_state="0" title="">27</td>
      <td class="ocal_state " data-ocal_state="0" title="">28</td>
      <td class="ocal_state " data-ocal_state="0" title="">29</td>
      <td class="ocal_state " data-ocal_state="0" title="">30</td>
    </tr>
    <tr>
      <td class="ocal_state " data-ocal_state="0" title="">31</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
      <td>&nbsp;</td>
    </tr>
  </tbody>
</table>
  <!-- pagination -->
<p class="ocal_pagination">
  <a class="ocal_button" href="/?ocal_year=2023&amp;ocal_month=7&amp;ocal_action=calendar">Today</a>
  <a class="ocal_button" href="/?ocal_year=2023&amp;ocal_month=8&amp;ocal_action=calendar">Next Month</a>
  <a class="ocal_button" href="/?ocal_year=2024&amp;ocal_month=7&amp;ocal_action=calendar">Next Year</a>
</p>
</div>
